Add more testing models for edge cases

Entenwilli opened the pull request:
This PR adds more testing models to test certain scenarios occurring in the analysis:
It provides testing models and tests for
- [x] Functionality in two different assemblies (Handling different node characteristics in the same SEFF Component)
- [x] Simple database component model (Load and store dataflow variables in data stores)
- [x] Cyclic dependency in execution order of action sequences (Logger.error and Exception)
- [x] Warning for databases that are never written to (Logger.warn)
- [x] Simple example of returning variable (Availability of RETURN dataflow variable, returning dataflow variables from SEFFs)
- [x] TODO: Composite components
- [x] TODO: All sources for node characteristics, i.e., resource container, user, assembly context, (composite components?)

This PR closes #23 and relates to previous work in #3 and #5

sebinside commented:
@Entenwilli I merged #22, please rebase this PR.

Entenwilli commented:
All commits should be correct now and the history cleaned up. 

As of now the analysis does not support Composite Components, because the Sequence Finder is broken for Composite Components. As i have not looked in depth at the part of the Sequence Finder which finds Called SEFFs, it may require more time to fix it. See `PCMQueryUtils:134`

Entenwilli commented:
Getting the label however should already be implemented, as all components that may have a label are present in the assembly context we create.
Only issue for now is, that we are unable to get the Effective Service Specification for the Called Component in Composite Components

sebinside commented:
Thank you for the update. We should discuss how we proceed with composite components in the next meeting with @majuwa

Entenwilli commented:
After some more digging around, i have found the issues in the analysis. I have fixed them, updated the tests and made sure everything still works after the modifications.

As this was the last TODO of this PR, this PR is now ready to be reviewed @sebinside, @majuwa

sebinside commented:
Thank you for finishing this PR and the update. Maybe we can together have a look at the test cases in our call next year? Besides that, the work on the [first milestone](https://github.com/PalladioSimulator/Palladio-Addons-DataFlowConfidentiality-Analysis/milestone/1) seems to be completed. Congratulations 🎉

sebinside commented on bundles/org.palladiosimulator.dataflow.confidentiality.analysis/src/org/palladiosimulator/dataflow/confidentiality/analysis/sequence/entity/pcm/PCMActionSequence.java:
```diff
@@ -147,20 +179,8 @@ public int compareTo(PCMActionSequence otherSequence) {
-				logger.error("Found incompatible set of Databases, Action Sequences depend on each other");
-				throw new IllegalStateException("Cylic loop of databases found in action sequences");
-			}
+			logger.error("Found incompatible set of Databases, Action Sequences depend on each other");
```
The error message should also contain the information of the *cyclic* dependency

sebinside commented on tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/ListAppender.java:
```diff
@@ -0,0 +1,37 @@
+import org.apache.log4j.Level;
+import org.apache.log4j.spi.LoggingEvent;
+
+public class ListAppender extends AppenderSkeleton {
```
Please add a short comment why this class is needed at all (debugging logging and so on)

sebinside commented on tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/ConstraintResultTest.java:
```diff
@@ -0,0 +1,192 @@
+    			.flatMap(it -> it.stream())
+    			.collect(Collectors.toList());
+    	
+    	assertEquals(constraintData.size(), results.size(), "Incorrect amount of violations found");
```
Here, but also at other parts of the code (e.g. in the data classes): The word `amount` usually specifies an uncountable number of items, while `count` refers to countable sets. See, e.g., this thread [here](https://ell.stackexchange.com/questions/3511/whats-the-difference-between-words-number-count-amount-and-quantity)

sebinside commented on tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/data/CharacteristicValueData.java:
```diff
@@ -0,0 +1,19 @@
+package org.palladiosimulator.dataflow.confidentiality.analysis.constraint.data;
+
+public class CharacteristicValueData {
```
Could this be a record?

sebinside commented on tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/data/ConstraintData.java:
```diff
@@ -0,0 +1,48 @@
+
+import de.uka.ipd.sdq.identifier.Identifier;
+
+public class ConstraintData {
```
Could this be a record?

Entenwilli replied:
I don't think it is necessary to use a record here, as the attributes of that class are only used internally

sebinside commented on tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/data/ConstraintViolations.java:
```diff
@@ -0,0 +1,57 @@
+import java.util.List;
+import java.util.Map;
+
+public class ConstraintViolations {
```
Please add a private constructor as this seems to be an utility class

sebinside commented on tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/ConstraintResultTest.java:
```diff
@@ -0,0 +1,192 @@
+    	Predicate<AbstractActionSequenceElement<?>> internationalOnlineShopConstraint = node -> internationalOnlineShopCondition(node);
+
+    	StandalonePCMDataFlowConfidentialtyAnalysis mutipleAssembliesAnalysis = 
+    			super.initializeAnalysis(Paths.get("models", "MultipleAssembliesTest", "default.usagemodel"), Paths.get("models", "MultipleAssembliesTest", "default.allocation"));
```
I'm not happy with the name `MultipleAssembliesTest`. I understand, that you allocated multiple dependencies on a single resource, but the name indicates that there are somehow multiple assemblies in the model (or something like that). However, I do not have a good name either. Maybe something like `MultipleAssembliesOneResourceTest` to clarify it?

Entenwilli replied:
I have renamed it to `OneAssemblyMutlipleResourceContainerTest` to indicate what this model is testing

sebinside commented on tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/data/ConstraintData.java:
```diff
@@ -0,0 +1,48 @@
+		return true;
+	}
+	
+	public int nodeCharacteristicsAmount() {
```
Same comment about the word `amount` that should be replaced with `count`

sebinside reviewed (changes requested):
Thank you for this PR, here is my first review. Unfortunately, I had problems opening the model files correctly, as expected due to #35. Still, I reviewed as much as possible. These are only minor remarks, after you updated the PR (and also please merge the main branch again), this can be merged. Thank you!
